### PR TITLE
Pad FP4 Marlin weights to valid thread tiles

### PR DESCRIPTION
**Pull request by mikekg**
## Purpose

FP4 Marlin repack and GEMM require rank-local N/K extents that match a supported thread tile family. A fixed K=128 pad works, but it overpads cases where the N128/K64 family is already valid.

Select the lower-overhead rank-local padded size from the two tile families that are valid for both small and large batches: N multiple 128 with K multiple 64, or N multiple 64 with K multiple 128.

Use the selected extents consistently when repacking weights and scales and when launching the Marlin GEMM. Pad the activation K dimension with zeros for the W4A16 path, slice padded output columns after GEMM, and pad bias before Marlin bias permutation when bias is present.

This padding is applied after tensor-parallel partitioning via output_size_per_partition and input_size_per_partition, so global checkpoint tensors are not expanded before TP slicing.

  ## Test Plan

  Validated the FP4 Marlin padding path with Gemma 4 NVFP4 models that
  previously exercised the Marlin FP4 repack/GEMM constraints.

  Commands used were equivalent to:

  ```bash
  vllm serve nvidia/Gemma-4-26B-A4B-NVFP4 \
    --quantization modelopt \
    --tensor-parallel-size 1 \
    --kv-cache-dtype bfloat16 \
    --trust-remote-code
```
  and the same setup for the Gemma 4 31B NVFP4 checkpoint.

  ## Test Result

  Gemma 4 26B NVFP4 (RTX 4090, L40S, A100, H100) and Gemma 4 31B NVFP4 (L40S, A100, H100) 
  both loaded and ran successfully with
  the padded FP4 Marlin path. The models completed request generation without
  the Marlin repack/GEMM tile-size failure.

  A100 BF16-KV vLLM serve sanity on Gemma-4-26B-A4B-NVFP4:

```text
  kv_cache_dtype=bfloat16
  tokenizer.class=GemmaTokenizer
  server ready after 205s
  RESPONSE identity: content='vllm-sanity-ok' tool_calls=[]
  RESPONSE math: content='42' tool_calls=[]
  RESPONSE tool_parser: content='' tool_calls=[{'id': 'chatcmpl-tool-
  baecf8d2b2fbc96c', 'type': 'function', 'function': {'name': 'get_weather',
  'arguments': '{\n  "city": "Paris"\n}'}}]
  PASS vllm_serve_sanity
```
The A100 validation stack also used:

https://github.com/vllm-project/vllm/pull/42610 for ModelOpt mixed precision support on Ampere.
https://github.com/vllm-project/vllm/pull/43330 for explicit native KV cache dtype in the Triton cache update path.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py [hunk 1]
```diff
@@ -179,6 +211,9 @@ def apply_fp4_marlin_linear(
 
         inputs, a_scales = marlin_quant_input(inputs, torch.float8_e4m3fn)
 
+    if input_dtype is None and padded_size_k != size_k:
+        inputs = torch.nn.functional.pad(inputs, (0, padded_size_k - size_k))
```
Comment (vadiklyutiy) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: Activation padding is gated on `input_dtype is None`  but the weight is repacked to `padded_size_k` unconditionally and the GEMM is launched with `size_k=padded_size_k`. For the MXFP4 + FP8-activation path  with a non-128-multiple K, the quantized activation still has `K=size_k` while the kernel reads `padded_size_k` - out-of-bounds/incorrect. NVFP4 W4A8 is rejected earlier, so this is the one reachable-but-untested case.

**Review by vadiklyutiy — COMMENTED** [review]
It would be good to add unit tests for new cases.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py (hunk 1 above)
Comment (mikekg) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py]: Thanks for review @vadiklyutiy !  Please check the updated PR in response!

**Comment by mikekg** [comment]
@vadiklyutiy 
> It would be good to add unit tests for new cases.

Done. Added to 
[tests/kernels/quantization/test_marlin_gemm.py](https://github.com/vllm-project/vllm/pull/43806/changes#diff-7fb180af137c03db26af18630176ddeee62844f966604e2164277adc456a10eb)

**Comment by vadiklyutiy** [comment]
@claude review

**Review by vadiklyutiy — APPROVED** [review]

**Comment by mikekg** [comment]
This was merged into omnibus Marlin padding patch https://github.com/vllm-project/vllm/pull/45295 and the issue is now closed
